### PR TITLE
Item management sup 10

### PR DESCRIPTION
**Pull request by FortuneAwaker**
Category

### CODE REVIEW

File: src/main/java/by/itech/lab/supplier/domain/Category.java [hunk 1]
```diff
@@ -15,48 +15,18 @@
...
-public class Category {
+public class Category implements BaseEntity {
 
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+    @Column(nullable = false, unique = true)
     private String category;
+    @Column(nullable = false)
     private Double taxRate;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/Category.java]: Maybe better to use BigDecimal?

File: src/main/java/by/itech/lab/supplier/dto/CategoryDto.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+@Data
+public class CategoryDto implements DefaultDto {
+
+    private Long id;
+    private String category;
+    private Double taxRate;
+    private Set<Item> items = new HashSet<>();
+    private boolean active;
+
+}
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/dto/CategoryDto.java]: no empty line at the end

File: src/main/java/by/itech/lab/supplier/domain/Category.java [hunk 3]
```diff
@@ -15,48 +15,18 @@
...
-    public Double getTaxRate() {
-        return taxRate;
-    }
-
-    public void setTaxRate(final Double taxRate) {
-        this.taxRate = taxRate;
-    }
-
-}
+}
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/Category.java]: no empty line at the end

**Review by IrynaZazyba — APPROVED** [review]

File: src/main/java/by/itech/lab/supplier/dto/mapper/BaseMapper.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package by.itech.lab.supplier.dto.mapper;
+
+import by.itech.lab.supplier.domain.BaseEntity;
+import by.itech.lab.supplier.dto.DefaultDto;
+
+public interface BaseMapper<E extends BaseEntity, D extends DefaultDto> {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/dto/mapper/BaseMapper.java]: What about a similar style over the whole app?

Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/dto/mapper/BaseMapper.java]: BaseEntity - BaseDto
DefaultEntity - DefaultDto
etc

File: src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    private Category fillCategory(final CategoryDto dto) {
+        Category category = new Category();
+        category.setCategory(dto.getCategory());
+        category.setTaxRate(dto.getTaxRate());
+        category.setItems(dto.getItems());
+        category.setActive(dto.isActive());
+        return category;
+    }
+}
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java]: line?

**Review by alesiahlushakova — APPROVED** [review]

File: src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public Category mapCategoryWithId(final CategoryDto dto) {
+        Category category = fillCategory(dto);
+        category.setId(dto.getId());
+        return category;
+    }
+
+    private Category fillCategory(final CategoryDto dto) {
+        Category category = new Category();
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java:33]: you've created builders but why you didnt use it?

File: src/main/java/by/itech/lab/supplier/repository/CategoryRepository.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    Optional<Category> findByCategory(final String categoryName);
+
+    @Modifying
+    @Query("update Category set active = false where id = :id")
+    void delete(@Param("id") Long id);
+
+    @Modifying
+    @Query("update Category set active = true where id = :id")
+    void activate(@Param("id") Long id);
+}
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/repository/CategoryRepository.java:23]: line?

File: src/main/java/by/itech/lab/supplier/repository/CategoryRepository.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+import java.util.Optional;
+
+@Repository
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    Optional<Category> findByCategory(final String categoryName);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/repository/CategoryRepository.java:14]: Did you test uniqueness?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    public CategoryDto findByCategory(final String categoryName) {
+        return categoryRepository.findByCategory(categoryName)
+          .map(mapper::map).orElse(null);
+    }
+
+    public CategoryDto save(final CategoryDto dto) {
+        Category category;
+        if (Objects.isNull(dto.getId())) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:34]: optional?

File: src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        dto.setId(entity.getId());
+        dto.setActive(entity.isActive());
+        dto.setCategory(entity.getCategory());
+        dto.setTaxRate(entity.getTaxRate());
+        dto.setItems(entity.getItems());
+        return dto;
+    }
+
+    public Category mapCategoryWithId(final CategoryDto dto) {
+        Category category = fillCategory(dto);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java:27]: you've created builders but why you didnt use it?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+@Service
+public class CategoryService {
+
+    private final CategoryRepository categoryRepository;
+
+    private final CategoryMapper mapper;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:19]: same issue with the style, repository - mapper or categoryRepository - categoryMapper

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    public CategoryService(final CategoryRepository categoryRepository,
+                           final CategoryMapper mapper) {
+        this.categoryRepository = categoryRepository;
+        this.mapper = mapper;
+    }
+
+    public CategoryDto findByCategory(final String categoryName) {
+        return categoryRepository.findByCategory(categoryName)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:28]: if you don do anything with null maybe you shouldnt use optional ?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        categoryRepository.delete(dto.getId());
+    }
+
+    @Transactional
+    public void activate(final CategoryDto dto) {
+        categoryRepository.activate(dto.getId());
+    }
+
+    public void deleteForever(final CategoryDto dto) {
+        categoryRepository.delete(mapper.mapCategoryWithId(dto));
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:63]: as I remember you've create special method for removing of entries (soft-removing), but here u use different one, why?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    public List<CategoryDto> readAll() {
+        return categoryRepository.findAll().stream()
+          .map(mapper::map).collect(Collectors.toList());
+    }
+
+    public CategoryDto findById(final Long id) {
+        return categoryRepository.findById(id).map(mapper::map)
+          .orElse(null);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:49]: what 'orElse' did to you? why its alone on the separate line?
// line splitting must be done in a similar style for the whole expression (or even better for the whole application), for example, you can split each expression:
```
categoryRepository.findById(id)
          .map(mapper::map)
          .orElse(null);
```
similarly will look logical split, as in the first line its retrieving and then you placed something for success case and after something for the failure case
or in one line, if this is a short expression that can be placed at one line (120 or 140 chars, based on code style agreement in the project)

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Transactional
+    public void delete(final CategoryDto dto) {
+        categoryRepository.delete(dto.getId());
+    }
+
+    @Transactional
+    public void activate(final CategoryDto dto) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:58]: maybe you dont need a link to the whole object in this method?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        categoryRepository.activate(dto.getId());
+    }
+
+    public void deleteForever(final CategoryDto dto) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:62]: maybe you dont need a link to the whole object in this method?

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+    public CategoryDto save(final CategoryDto dto) {
+        Category category;
+        if (Objects.isNull(dto.getId())) {
+            category = mapper.map(dto);
+        } else {
+            category = mapper.mapCategoryWithId(dto);
+        }
+        CategoryDto created = mapper.map(categoryRepository.save(category));
+        return created;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java]: why its separate line? are there some requirements for this?

File: src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml [hunk 18]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: SUP-10-1
+      author: alexander_popovich
+      changes:
+        - addColumn:
+            tableName: category
+            columns:
+              - column:
+                  name: is_active
```
Comment (svareyko) [inline src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml]: its not a good idea to place prefixes such 'is' in database column names, as there is a possible misunderstanding in the naming of fields in java (if use default mapping strategy - which is preferable)

File: src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+                  type: BOOLEAN
+                  defaultValue: true
+                  constraints:
+                    nullable: false
```
Comment (svareyko) [inline src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml]: line?

File: src/test/java/by/itech/lab/supplier/CategoryTest.java [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+        assert !readCategory.isActive();
+        readCategory = service.findById(readCategory.getId());
+        service.activate(category);
+        readCategory = service.findById(readCategory.getId());
+        assert readCategory.isActive();
+        assert service.readAll().contains(readCategory);
+        service.deleteForever(readCategory);
+        assert service.findById(readCategory.getId()) == null;
+    }
+}
```
Comment (svareyko) [inline src/test/java/by/itech/lab/supplier/CategoryTest.java]: line?

File: src/test/java/by/itech/lab/supplier/CategoryTest.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Test
+    void checkConnection() {
+        CategoryDto category = new CategoryDto();
+        category.setCategory("Test");
+        category.setTaxRate(20.0);
+        category.setActive(true);
+        category = service.save(category);
+        CategoryDto readCategory = service.findByCategory("Test");
+        assert category.equals(readCategory);
```
Comment (svareyko) [inline src/test/java/by/itech/lab/supplier/CategoryTest.java]: can you please split logically your checks? basically, each test contains 3 parts:
- what we have - given
- what we do - when
- what we expect - then

for example we're testing method which sum 2 variables:
```
// given
final Integer a = 25;
final Integer b = 25

// when
final Integer result = someService.sum(a, b);

// then
assertThat(result).isEqualTo(a + b);
```

**Review by svareyko — CHANGES_REQUESTED** [review]

File: src/main/java/by/itech/lab/supplier/dto/CategoryDto.java (hunk 2 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/dto/CategoryDto.java]: done

File: src/main/java/by/itech/lab/supplier/domain/Category.java (hunk 3 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/domain/Category.java]: done

File: src/main/java/by/itech/lab/supplier/dto/mapper/BaseMapper.java (hunk 4 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/dto/mapper/BaseMapper.java]: done

File: src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java (hunk 5 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/dto/mapper/CategoryMapper.java]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 17 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java]: done

File: src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml (hunk 18 above)
Comment (FortuneAwaker) [reply src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml]: done

File: src/main/java/by/itech/lab/supplier/domain/Category.java (hunk 1 above)
Comment (FortuneAwaker) [reply src/main/java/by/itech/lab/supplier/domain/Category.java]: done

File: src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml (hunk 19 above)
Comment (FortuneAwaker) [reply src/main/resources/db/changelog/1.0/2020-10-21-change-table-category.yaml]: done

File: src/test/java/by/itech/lab/supplier/CategoryTest.java (hunk 20 above)
Comment (FortuneAwaker) [reply src/test/java/by/itech/lab/supplier/CategoryTest.java]: done

File: src/test/java/by/itech/lab/supplier/CategoryTest.java (hunk 21 above)
Comment (FortuneAwaker) [reply src/test/java/by/itech/lab/supplier/CategoryTest.java]: done

File: src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java (hunk 13 above)
Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/service/impl/CategoryService.java:63]: its different method which not makes soft-remove, if you still didnt understood

**Pull request merged**